File: 19.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeNthFromEnd(self, head: Optional[ListNode], n: int) -> Optional[ListNode]:
        foo = ListNode(0, head)
        p, q = foo, foo
        for _ in range(n + 1):
            q = q.next
        while q is not None:
            p = p.next
            q = q.next
        p.next = p.next.next
        return foo.next

File: 19.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
struct Solution {
    ListNode* removeNthFromEnd(ListNode* head, int n) {
        ListNode foo(0, head);
        ListNode *p = &foo, *q = &foo;
        for (int i = 0; i <= n; ++i){
            q = q->next;
        }
        while (q != nullptr){
            p = p->next;
            q = q->next;
        }
        p->next = p->next->next;
        return foo.next;
    }
};
